routes/s3handler.js

```diff
@@ -0,0 +1,239 @@
+import express from 'express'
+import CryptoJS from 'crypto-js'
+import bodyParser from 'body-parser'
+import getObjectURLFromS3 from '../src/utils/s3'
+import faceUtils from '../src/utils/face++'
+
+const clientSecretKey = process.env.AWS_SECRET_ACCESS_KEY
+const expectedMinSize = 0
+const expectedMaxSize = 5000000
+const expectedBucket = 'uploadedphotostomatch'
+const expectedHostname = 'uploadedphotostomatch.s3.amazonaws.com'
+
+const router = express.Router()
+
+// Signs any requests.  Delegate to a more specific signer based on type of request.
+function signRequest(req, res) {
+    if (req.body.headers) {
+        signRestRequest(req, res);
+    }
+    else {
+        signPolicy(req, res);
+    }
+}
+
+// Signs multipart (chunked) requests.
+function signRestRequest(req, res) {
+    var version = req.query.v4 ? 4 : 2,
+        stringToSign = req.body.headers,
+        signature = version === 4 ? signV4RestRequest(stringToSign) : signV2RestRequest(stringToSign);
+
+    var jsonResponse = {
+        signature: signature
+    };
+
+    res.setHeader("Content-Type", "application/json");
+
+    if (isValidRestRequest(stringToSign, version)) {
+        res.end(JSON.stringify(jsonResponse));
+    }
+    else {
+        res.status(400);
+        res.end(JSON.stringify({invalid: true}));
+    }
+}
+
+function signV2RestRequest(headersStr) {
+    return getV2SignatureKey(clientSecretKey, headersStr);
+}
+
+function signV4RestRequest(headersStr) {
+    var matches = /.+\n.+\n(\d+)\/(.+)\/s3\/aws4_request\n([\s\S]+)/.exec(headersStr),
+        hashedCanonicalRequest = CryptoJS.SHA256(matches[3]),
+        stringToSign = headersStr.replace(/(.+s3\/aws4_request\n)[\s\S]+/, '$1' + hashedCanonicalRequest);
+
+    return getV4SignatureKey(clientSecretKey, matches[1], matches[2], "s3", stringToSign);
+}
+
+// Signs "simple" (non-chunked) upload requests.
+function signPolicy(req, res) {
+    var policy = req.body,
+        base64Policy = new Buffer(JSON.stringify(policy)).toString("base64"),
+        signature = req.query.v4 ? signV4Policy(policy, base64Policy) : signV2Policy(base64Policy);
+
+    var jsonResponse = {
+        policy: base64Policy,
+        signature: signature
+    };
+
+    res.setHeader("Content-Type", "application/json");
+
+    if (isPolicyValid(req.body)) {
+        res.end(JSON.stringify(jsonResponse));
+    }
+    else {
+        res.status(400);
+        res.end(JSON.stringify({invalid: true}));
+    }
+}
+
+function signV2Policy(base64Policy) {
+    return getV2SignatureKey(clientSecretKey, base64Policy);
+}
+
+function signV4Policy(policy, base64Policy) {
+    var conditions = policy.conditions,
+        credentialCondition;
+
+    for (var i = 0; i < conditions.length; i++) {
+        credentialCondition = conditions[i]["x-amz-credential"];
+        if (credentialCondition != null) {
+            break;
```

Comment:
What happens here if the first condition is null? Are you just trying to find the condition with a `x-amz-credential`?